### PR TITLE
snappi: add multi-ingress MACsec test architecture

### PR DESCRIPTION
**Pull request by varukann**
### Description of PR

Summary:

This PR adds Snappi MACsec support to the M2O fluctuating-lossless test using a two-ingress, one-egress topology.

Both ingress links use MACsec, while the shared egress link uses plain Ethernet and IPv4. The changes add MACsec interface configuration, device-endpoint mapping, egress-only tracking, PGID-based statistics processing, MACsec overhead-aware loss calculations, and cleanup handling.

The following files are included:

- `tests/common/snappi_tests/snappi_fixtures.py`
- `tests/common/snappi_tests/traffic_generation.py`
- `tests/snappi_tests/pfc/files/m2o_fluctuating_lossless_helper.py`
- `tests/snappi_tests/docs/snappi_macsec_two_ingress_one_egress_hld.md`

No additional Python package dependencies are introduced.

Fixes #<issue-number>

### Type of change

- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [ ] New Test case
    - [ ] Skipped for non-supported platforms
- [x] Test case improvement

### Back port request

- [ ] 202311
- [ ] 202405
- [ ] 202411
- [ ] 202505
- [ ] 202511
- [ ] 202512
- [ ] 202605

Tracking issue/work item for backport/cherry-pick request (GitHub issue or Microsoft ADO): N/A  
Failure type: N/A

### Tested branch

- [x] master
- [ ] 202311
- [ ] 202405
- [ ] 202411
- [ ] 202505
- [ ] 202511
- [ ] 202512
- [ ] 202605
- [ ] N/A

### Test result

- master:
  - Executed the M2O fluctuating-lossless test with `--snappi_macsec`.
  - Verified traffic from two MACsec ingress links to one plain IPv4 egress link.
  - Observed aggregate DUT drop: approximately 5.11%.
  - Expected aggregate DUT drop: approximately 5.93%.
  - Difference: approximately 0.81%, within the configured ±1% tolerance.
  - Verified MACsec PGID statistics for both ingress ports.
  - Verified that no debugging breakpoints remain.
  - Lint checks passed for the modified files.

#### What is the motivation for this PR?

The existing M2O fluctuating-lossless test used port-to-port Snappi flows and standard Snappi flow metric objects. This model does not provide the endpoint mapping and statistics required when MACsec is enabled.

In MACsec mode:

- Each ingress flow uses a MACsec device endpoint.
- The egress uses plain IPv4 device endpoints.
- Traffic from both ingress ports converges on one egress port.
- IxNetwork reports Flow Statistics rows instead of standard Snappi flow metric objects.
- PGIDs are allocated in blocks of seven for each ingress port.
- MACsec removes 32 bytes of overhead before traffic reaches the plain egress, changing the expected egress demand and loss.

The Snappi fixture, traffic-generation framework, and M2O verification logic therefore require MACsec-specific handling.

#### How did you do it?

The following changes were made in `snappi_fixtures.py`:

- Added MACsec setup for all ingress ports with `port_id >= 1`.
- Kept `port_id == 0` as the plain IPv4 egress port.
- Configured the DUT MACsec profile and enabled it on ingress interfaces.
- Created Snappi Ethernet, IPv4, SecY, secure-channel, and MKA objects.
- Created seven plain IPv4 egress tracking endpoints for every ingress port.
- Calculated the required egress prefix dynamically:
  - `/28` for one ingress and one egress.
  - `/27` for two ingress ports and one egress.
- Added egress-only tracking configuration.
- Added static ARP and route configuration for tracking endpoints.
- Tracked MACsec-enabled ports and configured DUT profiles for cleanup.
- Restored temporary interface addressing and removed MACsec configuration during cleanup.

The following changes were made in `traffic_generation.py`:

- Configured every MKA and StaticMacsec endpoint after applying the Snappi configuration.
- Set the correct DUT SCI MAC for each ingress MACsec endpoint.
- Disabled gratuitous ARP generation for MACsec endpoints.
- Installed ARP entries for all `7 × ingress_count` egress tracking endpoints.
- Added MACsec in-flight statistics logging.
- Filtered zero-transmit statistics rows.
- Included normalized priority information in statistics output.
- Normalized multi-ingress PGIDs using `PGID % 7`.
- Used Tx port, Rx port, and PGID to identify MACsec statistics rows.

The following changes were made in `m2o_fluctuating_lossless_helper.py`:

- Added MACsec device-to-device flow generation.
- Mapped each ingress source and priority to a distinct egress IPv4 endpoint.
- Added `flow_name_prio_map` handling for MACsec statistics.
- Added MACsec Flow Statistics verification using:
  - `PGID`
  - `Tx Port`
  - `Rx Port`
  - `Tx Frames`
  - `Rx Frames`
- Verified zero loss for test-flow priorities.
- Verified average background-flow loss against the calculated expectation.
- Added a 32-byte MACsec overhead adjustment to the DWRR demand calculation.
- Unified VOQ and non-VOQ DUT drop calculations:
  - VOQ platforms use combined ingress `rx_drp` counters.
  - Non-VOQ platforms use the egress `tx_drp` counter.
- Replaced the fixed aggregate-drop expectation with a dynamically calculated DWRR expectation.

#### How did you verify/test it?

The test was executed using the Snappi MACsec option:

```bash
pytest tests/snappi_tests/pfc/test_m2o_fluctuating_lossless.py \
    --testbed=<testbed> \
    --inventory=<inventory> \
    --snappi_macsec

### CODE REVIEW

**Comment by varukann** [comment]
@amitpawar12 @rawal01 @sdszhang @selldinesh for viz

**Comment by amitpawar12** [comment]
@vmittal-msft - this is MACSEC snappi pull-request for m2o fluctuating testcase.

Thanks,
-Amit

**Comment by sdszhang** [comment]
same changes as of https://github.com/sonic-net/sonic-mgmt/pull/23268,  only need to merge one of these two.

**Comment by sdszhang** [comment]
to add port-channel skip. ready in this week.

File: tests/common/snappi_tests/traffic_generation.py [hunk 1]
```diff
@@ -948,7 +956,7 @@ def run_traffic(duthost,
                 if int(metric['PGID']) in snappi_extra_params.flow_name_prio_map.values()
                 and metric['Tx Port'] == snappi_extra_params.base_flow_config["tx_port_name"]
             ]
-            if list(set(transmit_states)) == [0]:   # Issue encountered, workaround is != instead of ==
+            if list(set(transmit_states)) != [0]:   # Issue encountered, workaround is != instead of ==
```
Comment (vmittal-msft) [inline tests/common/snappi_tests/traffic_generation.py:959]: 🟠 **Inverted flow-stop condition.** `!= [0]` breaks the wait loop while `Tx Frame Rate` is still non-zero, and — conversely — when flows have actually stopped (all rates `0`) it falls into `else`, loops to `max_attempts`, and fails `pytest_assert(attempts < max_attempts)`. The `# workaround` comment and HLD §12.4 both call this out. It only passes today because the flows are fixed-duration and have already ended by the time this loop runs.

Also note `set_flow_transmit_state(api, "stop", …)` just above is gated `not ptype`, so in MACsec mode the data flows are never explicitly stopped here.

Suggested fix: use `== [0]`, also stop the data flows in `ptype` mode, and handle the empty-rows case (`transmit_states == []`) explicitly rather than letting it break the loop.

File: tests/snappi_tests/pfc/files/m2o_fluctuating_lossless_helper.py [hunk 2]
```diff
@@ -533,17 +617,8 @@ def __gen_data_flow(testbed_config,
...
-    UDP_PORT_START += no_of_streams
-    udp.src_port.increment.start = src_port
-    udp.src_port.increment.step = 1
-    udp.src_port.increment.count = no_of_streams
-
-    ipv4.src.value = tx_port_config.ip
-    ipv4.dst.value = gen_data_flow_dest_ip(rx_port_config.ip)
     ipv4.priority.choice = ipv4.priority.DSCP
-
+    flow_prio = sorted(flow_prio)
```
Comment (vmittal-msft) [inline tests/snappi_tests/pfc/files/m2o_fluctuating_lossless_helper.py:621]: 🟠 **`sorted(flow_prio)` desynchronizes classification from measurement.** This runs *after* the MACsec block already set `fp`, `flow_name_prio_map[flow.name]`, the rx-endpoint device index, and `eth.pfc_queue.value` from the **original** order — but the DSCP marking below uses the **sorted** order. In the test, `bg_prio_list = random.sample(...)` (test_m2o_fluctuating_lossless.py:69) is unsorted, so a background flow can be DSCP-marked as priority X (→ the DUT queue/DWRR weight that actually determines its loss) while its PGID endpoint, map label, and PFC-queue tag all reference priority Y.

The averaged background-loss check may still pass, but per-flow attribution is wrong and `pfc_queue` (unsorted) disagrees with DSCP (sorted). Please sort once at the very top of the function before any use, or drop the sort entirely so a single consistent ordering is used throughout.

File: tests/snappi_tests/pfc/files/m2o_fluctuating_lossless_helper.py [hunk 3]
```diff
@@ -553,10 +628,12 @@ def __gen_data_flow(testbed_config,
...
         ipv4.priority.dscp.phb.values = prio_dscp_map[flow_prio[3]]
     elif 'Test Flow 1 -> 0' in flow.name:
-        ipv4.priority.dscp.phb.values = [flow_prio[0]]
+        ipv4.priority.dscp.phb.values = prio_dscp_map[flow_prio[0]]
     elif 'Test Flow 2 -> 0' in flow.name:
-        ipv4.priority.dscp.phb.values = [flow_prio[1]]
-
+        ipv4.priority.dscp.phb.values = prio_dscp_map[flow_prio[1]]
+    if len(ipv4.priority.dscp.phb.values) > 1:
+        ipv4.priority.dscp.phb.values = ipv4.priority.dscp.phb.values[:1]
```
Comment (vmittal-msft) [inline tests/snappi_tests/pfc/files/m2o_fluctuating_lossless_helper.py:635]: 🟡 **Behavior change to the existing non-MACsec path.** This `[:1]` truncates the DSCP list for *all* flows — including background flows, which previously sent the full `prio_dscp_map[prio]` list. Combined with the Test-Flow change from `[flow_prio[0]]` to `prio_dscp_map[flow_prio[0]]` (lines 623/631), the non-MACsec m2o test's DSCP marking is now different from before.

The PR notes testing only with `--snappi_macsec`. Could you re-run the baseline (non-MACsec) `test_m2o_fluctuating_lossless.py` to confirm no regression, or gate these MACsec-specific changes behind the `ptype` branch? Also consider dropping the leftover per-flow `logger.info('{} {} {}'…)` debug line just below.

File: tests/common/snappi_tests/snappi_fixtures.py [hunk 4]
```diff
@@ -140,7 +141,7 @@ def __gen_mac(id):
     Returns:
         MAC address (string)
     """
-    return '00:{:02d}:22:33:44:01'.format(id)
+    return '00:{:02x}:22:33:44:01'.format(id)
```
Comment (vmittal-msft) [inline tests/common/snappi_tests/snappi_fixtures.py:144]: 🟡 **Shared helper — wider blast radius than MACsec.** `__gen_mac` is used by every topology config (also called around lines 212/320/419/1388/1437), not just the MACsec path. `:02d`→`:02x` produces identical output for `id < 10` but differs for `id >= 10`, so any testbed with ≥10 Snappi ports would get different generated MACs in non-MACsec configs too. If this is only needed for the MACsec endpoints, please scope it there; otherwise confirm it's safe for large non-MACsec setups.

File: tests/snappi_tests/pfc/files/m2o_fluctuating_lossless_helper.py [hunk 5]
```diff
@@ -578,9 +656,52 @@ def verify_m2o_fluctuating_lossless_result(rows,
...
+            if tx_frames == 0 or row['Rx Port'] != rx_port_name:
+                continue
+
+            # MACsec PGIDs use blocks of 7 per Tx port; modulo recovers the original priority.
+            prio = int(row['PGID']) % 7
+            rx_frames = int(row['Rx Frames'])
+            loss = 100.0 * (tx_frames - rx_frames) / tx_frames
+
+            if prio in test_prios:
+                pytest_assert(int(loss) == 0,
```
Comment (vmittal-msft) [inline tests/snappi_tests/pfc/files/m2o_fluctuating_lossless_helper.py:691]: ⚪ Two minor robustness points in this MACsec verify block:
1. `int(loss) == 0` truncates, so any loss in [0%, 1%) passes as "lossless". This matches the existing non-MACsec style, but for a strict lossless assertion consider an explicit epsilon.
2. Rows are classified purely via `int(row['PGID']) % 7 in test_prios/bg_prios`. If a priority is ever shared between test and background flows, `if prio in test_prios … elif prio in bg_prios` misclassifies them (test wins). Per HLD §9, prefer identifying a row by `(Tx Port, Rx Port, PGID)` and applying `% 7` only to recover the logical priority.

**Review by vmittal-msft — COMMENTED** [review]
Thanks for the MACsec architecture work and the detailed HLD — the multi-port bookkeeping, per-port unique object names, portchannel-member guard, and the `egress_duthost.facts['switch_type']` fix (base leaked `duthost` from a `for … in set` loop) are solid improvements.

I've left inline comments. Two are effectively blocking:
- **High (traffic_generation.py:959):** the flow-stop condition is inverted (`!= [0]`) — the code comment and HLD §12.4 both acknowledge it.
- **High (m2o_fluctuating_lossless_helper.py:621):** `sorted(flow_prio)` desyncs the DUT-side DSCP classification from the MACsec measurement/endpoint mapping, since `bg_prio_list` is `random.sample(...)`.

The rest are medium/low: a couple of DSCP changes that also affect the existing non-MACsec path (please re-validate the baseline run), the shared `__gen_mac` format change, and minor robustness notes in the verify block.

File: tests/common/snappi_tests/traffic_generation.py (hunk 1 above)
Comment (varukann) [reply tests/common/snappi_tests/traffic_generation.py:959]: we agree that the current` !=[0]` workaround is broad but the underlying problem is a known IxNetwork/Snappi MACsec traffic-state reporting issue tracked in [https://github.com/sonic-net/sonic-mgmt/issues/27899](url). we will address the stop condition and empty row handling there and keep this PR unchanged.

File: tests/snappi_tests/pfc/files/m2o_fluctuating_lossless_helper.py (hunk 2 above)
Comment (varukann) [reply tests/snappi_tests/pfc/files/m2o_fluctuating_lossless_helper.py]: for this issue, we removed the `sorted(flow_prio)` so the original priority order is now used consistently for the PGID endpoint, flow mapping,  PFC queue, and DSCP marking. this keeps traffic classification aligned with the corresponding measurement now.

File: tests/snappi_tests/pfc/files/m2o_fluctuating_lossless_helper.py (hunk 3 above)
Comment (varukann) [reply tests/snappi_tests/pfc/files/m2o_fluctuating_lossless_helper.py:646]: fixed by gating the DSCP-list truncation behind the MACsec ptype condition and restoring the original DSCP values for non-MACsec test flows. Non-MACsec background flows now retain their complete DSCP lists. the leftover per flow debug log was also removed.

**Comment by vmittal-msft** [comment]
@atul-nexthop @amitpawar12

File: tests/snappi_tests/pfc/files/m2o_fluctuating_lossless_helper.py (hunk 5 above)
Comment (varukann) [reply tests/snappi_tests/pfc/files/m2o_fluctuating_lossless_helper.py]: fixed both the comments, the lossless flows now require exact `tx_frames == rx_frames`, so fractional loss can no longer pass. statistics rows are classified using (Tx Port, Rx Port, PGID) with %7 used only to recover the logical priority. missing test-flow rows also fail explicitly.

File: tests/common/snappi_tests/snappi_fixtures.py (hunk 4 above)
Comment (varukann) [reply tests/common/snappi_tests/snappi_fixtures.py:144]: this change is intentional and MAC octets are hexadecimal, so :02x correctly supports the full 00-ff range. :02d becomes invalid for `IDs >= 100`. since these MACs are used consistently throughout the configuration hence changing their representation does not affect functionality.

**Review by vmittal-msft — APPROVED** [review]
Approving for master — thanks for the thorough follow-up. Verified all prior comments against `36fb573`:

- **`sorted(flow_prio)` desync (High): fixed.** The sort is gone; the rx device slot, PGID identity, `fp`, `flow_name_prio_map`, `pfc_queue`, and DSCP now all use the same `flow_prio[i]` order, so classification and measurement stay aligned. With `test_prio_list={3,4}` disjoint from `bg_prio_list ⊂ {0,1,2,5,6}`, the new `isdisjoint` identity guard is safe.
- **Non-MACsec DSCP path (Med): fixed.** Test-flow DSCP and the `[:1]` truncation are now gated behind `ptype`; the non-MACsec path is byte-equivalent to baseline, and the per-flow debug log was removed.
- **verify block (Low): fixed.** Test flows now require exact `tx_frames == rx_frames`, rows are keyed by `(Tx Port, Rx Port, PGID)` with `%7` only recovering the priority, and missing test-flow rows fail explicitly.
- **`__gen_mac` `:02x` (Med):** agreed — hex is correct for MAC octets and is generated/consumed consistently.
- **Inverted `!= [0]` flow-stop (High):** accepting the deferral to #27899 with the acknowledgment in code. Please make sure that issue is prioritized so the workaround (and empty-row handling) doesn't linger.

Non-blocking nits for a future cleanup: the `"--snappi_macsec" in sys.argv` sniffing is now repeated across several functions (a `snappi_extra_params` flag would be cleaner), and the flow-gen hardcodes exactly two ingress ports (`(2 - 1) * 7`) despite the "multi-ingress" framing.

Two soft asks before/at merge: confirm the pipeline check-runs are green, and that a non-MACsec baseline `test_m2o_fluctuating_lossless.py` run still passes (low risk now that the changes are gated). LGTM.
